compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -0,0 +1,128 @@
+use rustc_ast::ast;
+use rustc_ast::visit;
+use rustc_ast::visit::Visitor;
+use rustc_ast::EnumDef;
+use rustc_ast::ForeignMod;
+use rustc_ast_lowering::ResolverAstLowering;
+use rustc_middle::middle::privacy::AccessLevel;
+
+use crate::Resolver;
+
+pub struct PrivacyVisitor<'r, 'a> {
+    r: &'r mut Resolver<'a>,
+    prev_level: Option<AccessLevel>,
+}
+
+impl<'r, 'a> PrivacyVisitor<'r, 'a> {
+    pub fn new(r: &'r mut Resolver<'a>, root_level: Option<AccessLevel>) -> PrivacyVisitor<'r, 'a> {
+        PrivacyVisitor { r, prev_level: root_level }
+    }
+}
+
+impl<'r, 'ast> Visitor<'ast> for PrivacyVisitor<'ast, 'r> {
+    fn visit_item(&mut self, item: &'ast ast::Item) {
+        let inherited_item_level = match item.kind {
+            // FIXME Is this the correct behavior for those macros?
+            ast::ItemKind::MacCall(..) | ast::ItemKind::MacroDef(..) => None,
```

Comment:
`MacCall` should be impossible here, you can panic if it appears.
`MacroDef` is a regular item so it should be in the list below.